### PR TITLE
Issue52 featured fragment for both en and jp languages

### LINKED ISSUE #52: Featured Fragment for newsroom

### PR DESCRIPTION
**Pull request by meejain**
Please always provide the [GitHub issue(s)](../issues) your PR is for, as well as test URLs where your change can be observed (before and after):

Fixes #52 

Test URLs:
- Original: https://www.sunstar-foundation.org/newsroom
- Before: https://main--sunstar-foundation--hlxsites.hlx.live/en/fragments/featured-articles-newsroom
- After: https://issue52-featured-fragment--sunstar-foundation--hlxsites.hlx.page/en/fragments/featured-articles-newsroom
- Before: https://main--sunstar-foundation--hlxsites.hlx.live/fragments/featured-articles-newsroom
- After: https://issue52-featured-fragment--sunstar-foundation--hlxsites.hlx.page/fragments/featured-articles-newsroom

- [ ] New Blocks introduced in this PR
      If yes, please provide details below

Block Name    | Documentation
------------- | -------------
 For e.g. cards | [doc-link](https://main--sunstar-foundation--hlxsites.hlx.page/tools/sidekick/library.html?plugin=blocks&path=/sidekick/blocks/cards&index=0)

- [ ] New variations introduced in this PR

Variation Name    | Documentation
------------- | -------------
 For e.g. cards (grid)  | [doc-link](https://main--sunstar-foundation--hlxsites.hlx.page/tools/sidekick/library.html?plugin=blocks&path=/sidekick/blocks/cards&index=2)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -43,9 +43,9 @@
     "stylelint": "15.10.3",
     "stylelint-config-standard": "34.0.0"
   },
-
   "dependencies": {
     "fs-extra": "^11.1.1",
-    "jslinq": "^1.0.22"
+    "jslinq": "^1.0.22",
+    "lint": "^0.8.19"
```
Comment (dnbute) [inline package.json:49]: Why is this needed?

File: blocks/highlight/highlight.js [hunk 2]
```diff
@@ -1,5 +1,6 @@
 export default async function decorate(block) {
   let spotlight = document.createElement('div');
+  const hideSpotlight = (block.classList.contains('feed-newsroom'));
```
Comment (sdmcraft) [inline blocks/highlight/highlight.js:3]: How about just keeping the spotlight hidden through 
```
.block.feed-newsroom spotlight {
  display: none;
}
```

File: blocks/highlight/highlight.css [hunk 3]
```diff
@@ -10,7 +10,7 @@
   color: var(--color-text);
 }
 
-.block.highlight > * {
+.block.highlight:not(.featured) > * {
```
Comment (sdmcraft) [inline blocks/highlight/highlight.css:13]: How about 
```
.block.highlight.featured > * {
  //whatever
}
```

File: blocks/feed/feed.js [hunk 4]
```diff
@@ -139,7 +152,17 @@ export default async function decorate(block) {
...
     .toList();
   block.innerHTML = '';
-  const blockContents = resultParsers[blockType](results, blockCfg);
+  const meta = getMetadata('page-style');
+  if (meta === 'featured' && blockType === 'highlight') {
+    let locale = (getLanguage(
+      window.location.pathname,
+      false,
+    ));
+    if (locale !== 'en') locale = 'jp';
```
Comment (sdmcraft) [inline blocks/feed/feed.js:161]: Why do we need this? If the default locale is to be `jp` , it should be handled in `getLanguage`

File: blocks/feed/feed.js [hunk 5]
```diff
@@ -139,7 +152,17 @@ export default async function decorate(block) {
     .take(blockCfg.count ? parseInt(blockCfg.count, 10) : 4)
     .toList();
   block.innerHTML = '';
-  const blockContents = resultParsers[blockType](results, blockCfg);
+  const meta = getMetadata('page-style');
```
Comment (sdmcraft) [inline blocks/feed/feed.js:155]: I'd suggest avoiding use of `page-style` for any purpose other than page level CSS styles.

File: blocks/feed/feed.js [hunk 6]
```diff
@@ -100,11 +110,14 @@ function getMetadataNullable(key) {
...
  */
 export default async function decorate(block) {
+  let blockContents = 0;
+  let queryObj = 0;
   const blockCfg = readBlockConfig(block);
   const blockName = (blockCfg['block-type'] ?? 'cards').trim().toLowerCase();
   const blockType = (blockName.split('(')[0]).trim();
   const variation = (blockName.match(/\((.+)\)/) === null ? '' : blockName.match(/\((.+)\)/)[1]).trim();
-  const queryObj = await queryIndex(`${getLanguage()}-search`);
+  if (`${getLanguage()}` === 'ja') queryObj = await queryIndex('jp-search');
```
Comment (sdmcraft) [inline blocks/feed/feed.js:119]: `getLanguage` should take care of language defaults and `ja/jp` etc.

File: blocks/feed/feed.js [hunk 7]
```diff
@@ -76,14 +76,24 @@ const resultParsers = {
...
       if (cardImage) {
-        row.push(cardImage);
+        if (result.path) {
+          const pathImg = document.createElement('a');
+          pathImg.href = result.path;
+          pathImg.append(cardImage);
+          row.push(pathImg);
+        } else row.push(cardImage);
+      }
+      if (meta === 'featured') {
```
Comment (sdmcraft) [inline blocks/feed/feed.js:86]: Rather than deriving this from `meta` , it should be derived from the `result` . Basically show `FEATURED` if its a featured article.

File: package.json (hunk 1 above)
Comment (meejain) [reply package.json:49]: Seems like this was added while running "npm install" on my local machine to track the lint issues.

File: blocks/highlight/highlight.js (hunk 2 above)
Comment (meejain) [reply blocks/highlight/highlight.js:3]: feed-newsroom is a block which is not yet leveraged in this featured section. This PR is for the upper featured section which will be further as a fragment in newsroom page where to list all the cards - feed-newsroom block will be leveraged. Pl advise.

File: blocks/highlight/highlight.css (hunk 3 above)
Comment (meejain) [reply blocks/highlight/highlight.css:13]: image

Sure. How about leveraging the below then - 

.block.highlight > * {
  flex: 1;
}

.block.highlight.featured > div:first-child {
  flex: 1;
}

.block.highlight.featured > div:last-child {
  flex: 0;
}

File: blocks/feed/feed.js (hunk 4 above)
Comment (meejain) [reply blocks/feed/feed.js:161]: Yes I was getting ja as the output of getLanguage, that's why. Seems like Today Ashutosh has done some changes to make the getLanguage - - -> resulting in "ja". May be once merged to main, will remove this check

image

File: blocks/feed/feed.js (hunk 6 above)
Comment (meejain) [reply blocks/feed/feed.js:119]: Yes I was getting ja as the output of getLanguage, that's why. Seems like Today Ashutosh has done some changes to make the getLanguage - - -> resulting in "ja". May be once merged to main, will remove this check

File: blocks/feed/feed.js (hunk 5 above)
Comment (meejain) [reply blocks/feed/feed.js:155]: Yeah there seems to be a lot of CSS changes on this featured page and that's why put its CSS styles in featured.css & hence used page-style

Comment (meejain) [reply blocks/feed/feed.js:155]: Sure working on it

Comment (meejain) [reply blocks/feed/feed.js]: Fixed

**Comment by jindaliiita** [comment]
Why this PR contains **ja** to **jp** changes ?

File: package.json (hunk 1 above)
Comment (jindaliiita) [reply package.json:49]: Its not needed eslint is already there in dev dependency

File: blocks/feed/feed.js [hunk 8]
```diff
@@ -76,14 +76,24 @@ const resultParsers = {
         }
       });
       if (cardImage) {
-        row.push(cardImage);
+        if (result.path) {
+          const pathImg = document.createElement('a');
+          pathImg.href = result.path;
+          pathImg.append(cardImage);
+          row.push(pathImg);
+        } else row.push(cardImage);
```
Comment (jindaliiita) [inline blocks/feed/feed.js:84]: can you please add braces for the readability

File: blocks/feed/feed.js [hunk 9]
```diff
@@ -139,7 +152,15 @@ export default async function decorate(block) {
     .take(blockCfg.count ? parseInt(blockCfg.count, 10) : 4)
     .toList();
   block.innerHTML = '';
-  const blockContents = resultParsers[blockType](results, blockCfg);
+  if (block.classList.contains('featured')) {
```
Comment (sdmcraft) [inline blocks/feed/feed.js:155]: `featured` should be present in `blockCfg` already. Authors should specify it in block config instead of a variant.

File: blocks/highlight/highlight.js (hunk 2 above)
Comment (sdmcraft) [reply blocks/highlight/highlight.js:3]: If the spotlight is to be hidden, I think the authors should explicitly specify it as a block variant while authoring.

File: blocks/feed/feed.js [hunk 10]
```diff
@@ -76,14 +76,24 @@ const resultParsers = {
...
       if (cardImage) {
-        row.push(cardImage);
+        if (result.path) {
+          const pathImg = document.createElement('a');
+          pathImg.href = result.path;
+          pathImg.append(cardImage);
+          row.push(pathImg);
+        } else row.push(cardImage);
+      }
+      if (result.featured === 'true' && block.classList.contains('featured')) {
```
Comment (sdmcraft) [inline blocks/feed/feed.js:86]: `featured` must be read from `blockCfg`

File: blocks/feed/feed.js [hunk 11]
```diff
@@ -76,14 +76,24 @@ const resultParsers = {
...
+        const divFeatured = document.createElement('div');
+        divFeatured.innerHTML = `<h5>${featuredInnerText}</h5>`;
```
Comment (sdmcraft) [inline blocks/feed/feed.js:88]: Maybe just directly ewad it from `placeholders.featured;` instead of reading elsewhere and passing it here.

File: blocks/feed/feed.css [hunk 12]
```diff
@@ -1 +1 @@
-/* eslint no-empty: "info" */
\ No newline at end of file
+/* eslint no-empty: "info" */
```
[the remark this answers is not in the document]
Comment (meejain) [reply blocks/feed/feed.css:1]: Not sure how it was added. Removed it. Many Thanks

File: blocks/feed/feed.js (hunk 9 above)
Comment (meejain) [reply blocks/feed/feed.js:155]: This featured is not of the one in blockCfg. This featured is of the block name - "feed (featured)", since there are certain unique changes to be done for latest featured card & not to every featured card.

File: blocks/feed/feed.js (hunk 10 above)
Comment (meejain) [reply blocks/feed/feed.js:86]: This featured is not of the one in blockCfg. This featured is of the block name - "feed (featured)", since there are certain unique changes to be done for latest featured card & not to every featured card.

Comment (meejain) [reply blocks/feed/feed.js:86]: With regards to https://github.com/hlxsites/sunstar-foundation/pull/55#discussion_r1431000125 - this is all copied as it is from the hightlight.js from sunstar - https://github.com/hlxsites/sunstar/blob/main/blocks/highlight/highlight.js - just to make consistency across the blocks.

File: blocks/feed/feed.js (hunk 11 above)
Comment (meejain) [reply blocks/feed/feed.js:88]: If we put placeholders.featured in the hightlight section - then means we need to add the await action as well - 

const placeholders = await fetchPlaceholders(locale); which will further need to make the whole function async. Due to which its not displaying anything. Already tested that. That is why putting the above in an already existing main async function & then calling the Highlight block, accordingly.  

Please advise.

File: blocks/highlight/highlight.css (hunk 3 above)
Comment (sdmcraft) [reply blocks/highlight/highlight.css]: ok

File: blocks/feed/feed.js (hunk 9 above)
Comment (sdmcraft) [reply blocks/feed/feed.js:157]: ok, I feel we are overloading the highlight block with too many things here. Looking at the page, I am wondering if this should a plain columns block block. Is there a reason we need it to be a feed block?

File: blocks/feed/feed.js (hunk 11 above)
Comment (sdmcraft) [reply blocks/feed/feed.js:90]: Fetching placeholder from anywhere is equivalent. The `await` itself does not cause any delay, its what you do inside the `await` that may cause delay,

File: blocks/header/header.js [hunk 13]
```diff
@@ -98,7 +98,7 @@ const navDecorators = { 'nav-top': decorateTopNav, 'nav-middle': decorateMiddleN
 export default async function decorate(block) {
   // fetch nav content
   const navMeta = getMetadata('nav');
-  const navPath = navMeta || (getLanguage() === 'ja' ? '/nav' : `/${getLanguage()}/nav`);
+  const navPath = navMeta || (getLanguage() === 'jp' ? '/nav' : `/${getLanguage()}/nav`);
```
Comment (sdmcraft) [inline blocks/header/header.js:101]: Can you synch your PR with `main` ? That would likely remove all the `ja/jp` differences here.

**Comment by meejain** [comment]
Closing this PR since the changes are already taken care by https://github.com/hlxsites/sunstar-foundation/pull/69
